### PR TITLE
Perf tests for simple GPU math kernels

### PR DESCRIPTION
**Pull request by jabraham17**
Adds a new perf test for simple GPU math kernels. The test is setup to be easy to extend with more math functions, however to start this test is only setup for `tanh`, `sqrt`, and `fabs`.

This add a perf test to track some of the peculiarities from https://github.com/chapel-lang/chapel/issues/25178 and provides a way to further investigate this in the future

Tested that `start_test --performance --perflabel gpu-` works properly

[Reviewed by @ShreyasKhandekar]

### CODE REVIEW

File: test/gpu/native/studies/math/mathKernels.chpl [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+inline proc kernel(param function: string, const D, ref A) {
+  @assertOnGpu
+  @gpu.blockSize(256)
+  foreach i in D {
+    A[i] = callFunc(function, i);
+  }
+}
+inline proc callFunc(param function: string, v) {
+  select function {
+    when "tanhf" do return tanh(v:real(32));
```
Comment (e-kayrakli) [inline test/gpu/native/studies/math/mathKernels.chpl:49]: Can we make this repeatedly do `tanh` so as to hide the kernel launch latency by increasing the mathematical intensity?

Comment (jabraham17) [reply test/gpu/native/studies/math/mathKernels.chpl:49]: I did that initially as a little loop around `tanh` that sums before setting it in the array. However, this completely changed the performance characteristics and codegen

No Loop:
Chapel: 2.4 ms
nvcc: 1.8 ms
clang: 1.8 ms

With Loop (100 iterations)
Chapel: 18 ms
nvcc: 40 ms
clang: 22 ms

Comment (e-kayrakli) [reply test/gpu/native/studies/math/mathKernels.chpl:49]: OK, I can't stop thinking about it now :)

What happens if you add a loop with 1 trip by default but that can be controlled with a `config const`?

Comment (jabraham17) [reply test/gpu/native/studies/math/mathKernels.chpl:49]: I do not know how to explain this.

Before, I had a `config const iterations` in the chapel code and `#define iterations` in the cuda code. This time, I did a single `config const iterations` that gets threaded through to the cuda calls.

With iterations=1, the Chapel code improves to match the cuda code, and the cuda code has no change. With iterations of 100, I get the following:

Chapel: 72ms
nvcc: 24ms
clang: 7ms

Comment (jabraham17) [reply test/gpu/native/studies/math/mathKernels.chpl:65]: This is very very weird, but for the sake of commiting this code I have just added the param `useInnerLoop`, along with a note about the weird loop behavior

Comment (e-kayrakli) [reply test/gpu/native/studies/math/mathKernels.chpl:65]: > This is very very weird

Stop trying to entice me!

**Review by ShreyasKhandekar — APPROVED** [review]
This looks good to me! Thanks for adding these tests and graphs!

As a side note:
There seems to be a lot of wiring needed to add tests to compare against CUDA, I wonder if our testing system could be expanded to allow us to easily test against CUDA or HIP versions for GPU code analogous to how we can [compare against C versions](https://chapel-lang.org/docs/main/developer/bestPractices/TestSystem.html#comparing-to-a-c-version)
It would probably help us add more tests like this (which, AFAIK, we would like to do) in the future with less initial effort and less maintenance.

**Comment by jabraham17** [comment]
> There seems to be a lot of wiring needed to add tests to compare against CUDA, I wonder if our testing system could be expanded to allow us to easily test against CUDA or HIP versions for GPU code analogous to how we can [compare against C versions](https://chapel-lang.org/docs/main/developer/bestPractices/TestSystem.html#comparing-to-a-c-version)
It would probably help us add more tests like this (which, AFAIK, we would like to do) in the future with less initial effort and less maintenance.

Adding a `.test.cu` to our test system sounds like a great idea! Rather than implementing it here, I think its probably worth its own issue.

Something that does come to mind in that approach is how to select the CUDA/HIP compiler? In this perf test I have two versions, one compiled with `nvcc` and one with `clang`. Would we need `.test.cu` and `.test.cpp` (with `-x cuda` in a `.compopts`)? Or some other more "elegant" solution

**Comment by ShreyasKhandekar** [comment]
Yeah, I definitely didn't mean you should implement it in this effort, I was just thinking out loud. I will spin off an issue to discuss this further and link it here

**Comment by ShreyasKhandekar** [comment]
https://github.com/chapel-lang/chapel/issues/25671

**Pull request merged**